crates/accelerate/src/synthesis/clifford/greedy_synthesis.rs

```diff
@@ -0,0 +1,707 @@
+            ));
+            self.symplectic_matrix.prepend_cx(min_qubit, qubit_b);
+
+            gate_seq.push((
+                StandardGate::HGate,
+                smallvec![],
+                smallvec![Qubit(qubit_b as u32)],
+            ));
+            self.symplectic_matrix.prepend_h(qubit_b);
+
+            gate_seq.push((
+                StandardGate::CXGate,
+                smallvec![],
+                smallvec![Qubit(qubit_b as u32), Qubit(min_qubit as u32)],
+            ));
+            self.symplectic_matrix.prepend_cx(qubit_b, min_qubit);
+        }
+
+        let a_len: usize = (a_qubits.len() - 1) / 2;
+        if a_len > 0 {
+            a_qubits.swap_remove(&min_qubit);
+        }
+
+        for qubit in 0..a_len {
+            gate_seq.push((
+                StandardGate::CXGate,
+                smallvec![],
+                smallvec![
+                    Qubit(a_qubits[2 * qubit + 1] as u32),
+                    Qubit(a_qubits[2 * qubit] as u32)
+                ],
+            ));
+            self.symplectic_matrix
+                .prepend_cx(a_qubits[2 * qubit + 1], a_qubits[2 * qubit]);
+
+            gate_seq.push((
+                StandardGate::CXGate,
+                smallvec![],
+                smallvec![Qubit(a_qubits[2 * qubit] as u32), Qubit(min_qubit as u32)],
+            ));
+            self.symplectic_matrix
+                .prepend_cx(a_qubits[2 * qubit], min_qubit);
+
+            gate_seq.push((
+                StandardGate::CXGate,
+                smallvec![],
+                smallvec![
+                    Qubit(min_qubit as u32),
+                    Qubit(a_qubits[2 * qubit + 1] as u32)
+                ],
+            ));
+            self.symplectic_matrix
+                .prepend_cx(min_qubit, a_qubits[2 * qubit + 1]);
+        }
+
+        Ok(())
+    }
+
+    /// The main synthesis function.
+    pub(crate) fn run(&mut self) -> Result<(usize, CliffordGatesVec), String> {
+        let mut clifford_gates = CliffordGatesVec::new();
+
+        while !self.unprocessed_qubits.is_empty() {
+            let costs: Vec<(usize, usize)> = self
+                .unprocessed_qubits
+                .iter()
+                .map(|q| self.compute_cost(*q).map(|cost| (cost, *q)))
+                .collect::<Result<Vec<_>, _>>()?;
+
+            let min_cost_qubit = costs.iter().min_by_key(|(cost, _)| cost).unwrap().1;
+
+            self.decouple_qubit(&mut clifford_gates, min_cost_qubit)?;
+
+            self.unprocessed_qubits.swap_remove(&min_cost_qubit);
+        }
+
+        adjust_final_pauli_gates(&mut clifford_gates, self.tableau, self.num_qubits)?;
+
+        Ok((self.num_qubits, clifford_gates))
+    }
+}
+
+/// Given a sequence of Clifford gates that correctly implements the symplectic matrix
+/// of the target clifford tableau, adds the Pauli gates to also match the phase of
+/// the tableau.
+fn adjust_final_pauli_gates(
```

Comment:
this function could also be useful to synthesize a Clifford for LNN later

Reply:
Sure! We will need to reorganize the code structure a bit once more Clifford synthesis algorithms are implemented.

Reply:
done in [bbc6ceb](https://github.com/Qiskit/qiskit/pull/12601/commits/bbc6ceb158cd537dfe5f6909932132a75e51029b)